plugins/inputs/modbus/request.go

```diff
@@ -9,24 +12,168 @@ type request struct {
+	var l uint64
+	for _, r := range requests {
+		l += uint64(r.length)
+	}
+	return l
+}
+
+// Only split too-large groups, but ignore all optimization potential
+func splitMaxBatchSize(g request, maxBatchSize uint16) []request {
+	var requests []request
+
+	idx := 0
+	for start := g.address; start < g.address+g.length; {
+		current := request{
+			fields:  []field{},
+			address: start,
+		}
+		for _, f := range g.fields[idx:] {
+			// End of field still fits into the batch
+			if f.address+f.length <= start+maxBatchSize {
+				current.fields = append(current.fields, f)
+				idx++
+			}
+		}
+
+		end := start + maxBatchSize
+		if end > g.address+g.length {
+			end = g.address + g.length
+		}
+		if idx >= len(g.fields) || g.fields[idx].address >= end {
+			current.length = end - start
+		} else {
+			current.length = g.fields[idx].address - start
+		}
+		start = end
+
+		if len(current.fields) > 0 {
+			requests = append(requests, current)
+		}
+	}
+
+	return requests
+}
+
+func shrinkGroup(g request, maxBatchSize uint16) []request {
+	var requests []request
+	var current request
+
+	for _, f := range g.fields {
+		// Just add the field and update length if we are still
+		// within the maximum batch-size
+		if current.length > 0 && f.address+f.length <= current.address+maxBatchSize {
+			current.fields = append(current.fields, f)
+			current.length = f.address - current.address + f.length
+			continue
+		}
+
+		// Ignore completely empty requests
+		if len(current.fields) > 0 {
+			requests = append(requests, current)
+		}
+
+		// Create a new request
+		current = request{
+			fields:  []field{f},
+			address: f.address,
+			length:  f.length,
+		}
+	}
+	if len(current.fields) > 0 {
+		requests = append(requests, current)
+	}
+
+	return requests
+}
+
+func optimitzeGroup(g request, maxBatchSize uint16) []request {
+	if len(g.fields) == 0 {
+		return nil
+	}
+
+	requests := shrinkGroup(g, maxBatchSize)
+	length := countRegisters(requests)
+
+	for i := 1; i < len(g.fields)-1; i++ {
+		// Always keep consecutive fields as they are known to be optimal
+		if g.fields[i-1].address+g.fields[i-1].length == g.fields[i].address {
+			continue
+		}
+
+		// Perform the split and check if it is better
+		// Note: This involves recursive optimization of the right side of the split.
+		current := shrinkGroup(request{fields: g.fields[:i]}, maxBatchSize)
+		current = append(current, optimitzeGroup(request{fields: g.fields[i:]}, maxBatchSize)...)
+		currentLength := countRegisters(current)
+
+		// Do not allow for more requests
+		if len(current) > len(requests) {
+			continue
+		}
+		// Try to reduce the number of registers we are trying to access
+		if currentLength >= length {
+			continue
+		}
+
+		// We found a better solution
+		requests = current
+		length = currentLength
+	}
+
+	return requests
+}
+
+func optimitzeGroupWithinLimits(fields []field, maxBatchSize uint16, maxExtraRegisters uint16) []request {
+	fields = removeAllOmitted(fields)
```

Comment:
You do not need to do this as the completely empty fields are removed _before_ calling the optimization!